sql/grants.sql

```diff
@@ -0,0 +1,181 @@
+        SELECT 'comment' UNION
+        SELECT 'emailtemplate' UNION
+        SELECT 'geolocation' UNION
+        SELECT 'interfacemessage' UNION
+        SELECT 'log' UNION
+        SELECT 'rdnscache' UNION
+        SELECT 'request' UNION
+        SELECT 'user' UNION
+        SELECT 'welcometemplate' UNION
+        SELECT 'xfftrustcache' FROM DUAL WHERE privmode = 'maintenance';
+
+    DECLARE cur_delete CURSOR FOR
+        SELECT 'antispoofcache' FROM DUAL WHERE privmode = 'maintenance' UNION
+        SELECT 'geolocation' FROM DUAL WHERE privmode = 'maintenance' UNION
+        SELECT 'rdnscache' FROM DUAL WHERE privmode = 'maintenance' UNION
+        SELECT 'xfftrustcache' FROM DUAL WHERE privmode = 'maintenance';
+
+    -- --------------------------------------------------------------------------------------------------
+    -- These two cursors find all existing mysql grants on this schema and systematically revoke them.
+    DECLARE cur_revschema CURSOR FOR
+        SELECT CONCAT('REVOKE ALL PRIVILEGES ON ', table_schema, '.* FROM ', grantee) vsql
+        FROM information_schema.schema_privileges
+        WHERE grantee = targetuser
+        GROUP BY table_schema, grantee;
+
+    DECLARE cur_revtable CURSOR FOR
+        SELECT CONCAT('REVOKE ALL PRIVILEGES ON ', table_schema, '.', table_name, ' FROM ', grantee) vsql
+        FROM information_schema.table_privileges
+        WHERE grantee = targetuser AND table_schema = schemaname
+        GROUP BY grantee, table_schema, table_name;
+
+    -- Continue handler for when a cursor fetch retrieves no data.
+    DECLARE CONTINUE HANDLER FOR NOT FOUND SET cursor_done = 1;
+
+    IF privmode NOT IN ('web', 'maintenance') THEN
+        SIGNAL SQLSTATE '45000' SET message_text = 'The privmode parameter should be either web or maintenance';
+    END IF;
+
+    IF targetuser NOT LIKE '\'%\'@\'%\'' THEN
+        SIGNAL SQLSTATE '45000' SET message_text = 'Please specify user in the standard MySQL format of \'user\'@\'host\'.';
+    END IF;
+
+    -- Revoke everything to begin with - looping over result sets from the revocation
+    -- cursors and running the dynamic sql; firstly schema-wide, then table-specific.
+    SET cursor_done = 0;
+    OPEN cur_revschema;
+    revschema: REPEAT
+        FETCH cur_revschema INTO vsql;
+        IF cursor_done THEN
+            LEAVE revschema;
+        END IF;
+        SET @sqlText = vsql;
+        PREPARE statement FROM @sqlText;
+        EXECUTE statement;
+        DEALLOCATE PREPARE statement;
+    UNTIL cursor_done END REPEAT revschema;
+    CLOSE cur_revschema;
+
+    SET cursor_done = 0;
+    OPEN cur_revtable;
+    revtable: REPEAT
+        FETCH cur_revtable INTO vsql;
+        IF cursor_done THEN
+            LEAVE revtable;
+        END IF;
+        SET @sqlText = vsql;
+        PREPARE statement FROM @sqlText;
+        EXECUTE statement;
+        DEALLOCATE PREPARE statement;
+    UNTIL cursor_done END REPEAT revtable;
+    CLOSE cur_revtable;
+
```

Comment:
Actually, most of that is just MySQL. PL/SQL is so much nicer.

Like, actually, not kidding:

```sql
CREATE PROCEDURE foo ( 
    snork IN VARCHAR2
  , blurble IN NUMBER 
)
IS
  CURSOR bar IS SELECT 'REVOKE ALL....' AS vsql FROM baz;

  rec bar%ROWTYPE;
BEGIN
  
  FOR rec IN bar LOOP
    EXECUTE IMMEDIATE rec.vsql;
  END LOOP;
  
END;
```